pkg/inventory/db.go

```diff
@@ -43,183 +45,187 @@ const (
+	deviceStore map[string]*resourceapi.Device
+
+	rateLimiter *rate.Limiter
+	// syncCh is a channel that signals a request for an immediate device
+	// discovery (rateLimiter is NOT ignored.)
+	syncCh        chan any
 	notifications chan []resourceapi.Device
 	hasDevices    bool
 }
 
 func New() *DB {
 	return &DB{
 		rateLimiter:   rate.NewLimiter(rate.Every(minInterval), 1),
-		podStore:      map[int]string{},
-		podNsStore:    map[string]string{},
+		syncCh:        make(chan any, 1),
+		netNsForPod:   map[string]string{},
 		notifications: make(chan []resourceapi.Device),
 	}
 }
 
-func (db *DB) AddPodNetns(pod string, netnsPath string) {
-	db.mu.Lock()
-	defer db.mu.Unlock()
-	ns, err := netns.GetFromPath(netnsPath)
-	if err != nil {
-		klog.Infof("fail to get pod %s network namespace %s handle: %v", pod, netnsPath, err)
-		return
-	}
-	defer ns.Close()
-	id, err := netlink.GetNetNsIdByFd(int(ns))
-	if err != nil {
-		klog.Infof("fail to get pod %s network namespace %s netnsid: %v", pod, netnsPath, err)
-		return
-	}
-	db.podStore[id] = pod
-	db.podNsStore[pod] = netnsPath
-}
-
-func (db *DB) RemovePodNetns(pod string) {
-	db.mu.Lock()
-	defer db.mu.Unlock()
-	delete(db.podNsStore, pod)
-	for k, v := range db.podStore {
-		if v == pod {
-			delete(db.podStore, k)
-			return
-		}
-	}
-}
-
-// GetPodName allows to get the Pod name from the namespace Id
-// that comes in the link id from the veth pair interface
-func (db *DB) GetPodName(netnsid int) string {
-	db.mu.RLock()
-	defer db.mu.RUnlock()
-	return db.podStore[netnsid]
-}
-
-// GetPodNamespace allows to get the Pod network namespace
-func (db *DB) GetPodNamespace(pod string) string {
-	db.mu.RLock()
-	defer db.mu.RUnlock()
-	return db.podNsStore[pod]
-}
-
 func (db *DB) Run(ctx context.Context) error {
 	defer close(db.notifications)
 
-	nlHandle, err := netlink.NewHandle()
-	if err != nil {
-		return fmt.Errorf("error creating netlink handle %v", err)
-	}
-	// Resources are published periodically or if there is a netlink notification
-	// indicating a new interfaces was added or changed
-	nlChannel := make(chan netlink.LinkUpdate)
-	doneCh := make(chan struct{})
-	defer close(doneCh)
-	if err := netlink.LinkSubscribe(nlChannel, doneCh); err != nil {
-		klog.Error(err, "error subscribing to netlink interfaces, only syncing periodically", "interval", maxInterval.String())
-	}
-
 	// Obtain data that will not change after the startup
 	db.instance = getInstanceProperties(ctx)
-	// TODO: it is not common but may happen in edge cases that the default gateway changes
-	// revisit once we have more evidence this can be a potential problem or break some use
-	// cases.
-	gwInterfaces := getDefaultGwInterfaces()
 
 	for {
 		err := db.rateLimiter.Wait(ctx)
 		if err != nil {
 			klog.Error(err, "unexpected rate limited error trying to get system interfaces")
 		}
 
-		devices := []resourceapi.Device{}
-		ifaces, err := nlHandle.LinkList()
```

Comment:
this logic remains

Reply:
or we just use sys/class/net but we keep the existing behavior working, just we have two source of truth now, the sys/devices and (sys/class/net or netlink)